src/main/java/io/cdap/plugin/sap/ODataEntryToRecordTransformer.java

```diff
@@ -0,0 +1,206 @@
+import java.util.Calendar;
+import java.util.GregorianCalendar;
+import java.util.UUID;
+import java.util.concurrent.TimeUnit;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+/**
+ * Transforms {@link ODataEntry} to {@link StructuredRecord}.
+ */
+public class ODataEntryToRecordTransformer {
+
+  private final Schema schema;
+
+  public ODataEntryToRecordTransformer(Schema schema) {
+    this.schema = schema;
+  }
+
+  /**
+   * Transforms given {@link ODataEntry} to {@link StructuredRecord}.
+   *
+   * @param oDataEntry ODataEntry to be transformed.
+   * @return {@link StructuredRecord} that corresponds to the given {@link ODataEntry}.
+   */
+  public StructuredRecord transform(ODataEntry oDataEntry) {
+    StructuredRecord.Builder builder = StructuredRecord.builder(schema);
+    for (Schema.Field field : schema.getFields()) {
+      Schema nonNullableSchema = field.getSchema().isNullable() ?
+        field.getSchema().getNonNullable() : field.getSchema();
+      String fieldName = field.getName();
+      Object value = oDataEntry.getProperties().get(fieldName);
+      builder.set(fieldName, extractValue(fieldName, value, nonNullableSchema));
+    }
+    return builder.build();
+  }
+
+  /**
+   * Extract value of EDM types according to the provided schema. Some of the EDM types can be represented by multiple
+   * Java types. For more information see:
+   * <a href="https://olingo.apache.org/javadoc/odata2/org/apache/olingo/odata2/api/edm/EdmSimpleType.html">
+   * EdmSimpleType
+   * </a>
+   */
+  private Object extractValue(String fieldName, Object value, Schema schema) {
+    if (value == null) {
+      return null;
+    }
+
+    Schema.LogicalType fieldLogicalType = schema.getLogicalType();
+    if (fieldLogicalType != null) {
+      switch (fieldLogicalType) {
+        case TIMESTAMP_MILLIS:
+          // Edm.DateTime
+          ensureTypeValid(fieldName, value, GregorianCalendar.class);
+          return extractTimestampMillis((GregorianCalendar) value);
+        case TIMESTAMP_MICROS:
+          // Edm.DateTime
+          ensureTypeValid(fieldName, value, GregorianCalendar.class);
+          return extractTimestampMicros((GregorianCalendar) value);
+        case TIME_MILLIS:
+          // Edm.Time
+          ensureTypeValid(fieldName, value, GregorianCalendar.class);
+          return extractTimeMillis((GregorianCalendar) value);
+        case TIME_MICROS:
+          // Edm.Time
+          ensureTypeValid(fieldName, value, GregorianCalendar.class);
+          return extractTimeMicros((GregorianCalendar) value);
+        case DECIMAL:
+          ensureTypeValid(fieldName, value, BigDecimal.class);
+          return extractDecimal(fieldName, (BigDecimal) value, schema);
+        default:
+          throw new UnexpectedFormatException(String.format("Field '%s' is of unsupported type '%s'", fieldName,
+                                                            fieldLogicalType.name().toLowerCase()));
```

Comment:
in other places we've used the token rather than the enum name

Reply:
Changed to use token.